src/python/controller/controller.py

```
@@ -608,7 +711,15 @@ def __propagate_exceptions(self):
         Propagate any exceptions from child processes/threads to this thread
         :return:
         """
-        self.__lftp.raise_pending_error()
+        try:
+            self.__lftp.raise_pending_error()
+        except LftpError as e:
+            error_str = str(e)
+            # Permanent errors (bad credentials, access denied) must propagate
+            permanent_patterns = ["Login failed", "Access failed"]
+            if any(p in error_str for p in permanent_patterns):
+                raise AppError(error_str)
```

Ruff (0.15.1)

[warning] Within an `except` clause, raise exceptions with `raise ... from err` or `raise ... from None` to distinguish them from errors in exception handling

(B904)